Support hyperspace

Frando opened the pull request:
This PR adds support for hyperspace.

It adds a new option `backend` that defaults to `native` (equals current SDK - run corestore, corestore-swarm-networker, etc). If set to `hyperspace`, or if `opts.hyperspaceClient` is present, the SDK instead connects to the hyperspace daemon, and exposes the same feature set via the hyperspace client.

All tests pass! :rocket: :smile:
(with the caveat that the `persist` option is silently ignored for the hyperspace backend)

There's a few TODOs:

### In-memory hypercores/hyperdrives

The one feature difference is that currently the dat-sdk supports in-memory hypercores/hyperdrives. This has no equivalent at the moment with hyperspace (it doesn't support specifying the storage for individual hypercores or hyperdrives). I see 4 possible ways to deal with this:
- a) Just ignore the `persist` option if on the hyperspace backend (this is what happens right now) and document properly
- b) Throw an error if creating cores with `persist: true` on the hyperspace backend. This is not super easy because core creation is a sync operation, where throwing exceptions is oftenly not expected, but if properly documented could work out too.
- c) Remove support for non-persisting cores from the SDK. Even on the native backends it's a bit hacky atm. Would work for me. 
- d) Push for proper support for non-persisting cores in upstream corestore and hyperspace. Not sure how feasable this is, it would add quite some complexity likely.

### Seperation of imports

To keep things simple while hacking on this so far I left everything in the index.js. We'll want to seperate the imports though, because many packages are needed for only either backend. There's a few ways to do that (e.g. leveljs does similar things with switching between leveldown - the native node module - and level-js). I can add a commit that just moves the backends and their imports into individual files.

RangerMauve commented on index.js:
```diff
@@ -85,8 +81,213 @@ async function SDK ({
+    makeHypercore
+  }
+
+  async function deriveSecret (namespace, name) {
```
This is a pretty important feature to me. 😅 Any chance we could get it in upstream? Might be good to add a note to the README about this change.

RangerMauve commented on package.json:
```diff
@@ -71,9 +72,11 @@
     "random-access-file": "^2.1.3",
     "tape": "^4.10.2",
     "tmp": "^0.1.0",
+    "tmp-promise": "^3.0.2",
```
Is this needed?

RangerMauve reviewed (changes requested):
Thank you for taking the time to put this together @Frando !

I like how the API surface isn't changing too much.

Would you mind adding some docs to the README that mention hyperspace and the new options it brings into the SDK constructor?

Regarding the persist flag, I think I'm okay with getting rid of the per-hypercore and per-hyperdrive persist. I've found that having a global persist:false has been my main use case lately and the per-core function is super hacky and fragile.

I think @calm-rad was using the per-drive persist flag in Gateway, though. Rad: Is it particularly important for your use case?

The lack of deriveSecret is going to be a bit of a pain, but I'm not sure how reasonable it'd be to get it added to corestore as a public API. 🤔

Does the hyperspace client work in browsers? Or is it just a node.js thing like the grpc API the hyperdrive daemon used? If it doesn't work in browsers, I feel like it might be useful to extract it into a separate module and keep the backend code you added for being able to switch it out.

Overall this is pretty exciting, and I could see it being handy for a lot of use cases. 😁

Frando commented:
I added two commits that remove the `persist` option for individual drives and cores, and that split the backends into two files.

Some answers to your questions:

- `@hyperspace/client` currently only works in Node, with a hyperspace server running on the same host. It connects to the server with a binary protocol ([hrpc](https://github.com/mafintosh/hrpc) over a Unix domain socket (or named pipe on windows). The hrpc protocol itself is really simple and should just work in browsers, however for that the hrpc connection would have to be proxied over Websockets. This does not exist yet. The proxying itself should be straightforward, however we'd have to think about authorization (currently, a hyperspace RPC connection allows you to write to all writable cores without any authentication).

- To support `deriveSecret`, we can either look into getting it added upstream, or try to directly read the masterKey from disc (will only work if in NodeJS)

This leaves, IMO:

- [ ] Decide on package structure - one, two, or three packages? `dat-sdk-native`, `dat-sdk-hyperspace` plus `dat-sdk` that selects one of them based on options? Or keep the native code `dat-sdk` and add `dat-sdk-hyperspace`? Or keep it all here, and make sure that the browser bundler currently only includes the native backend?
- [ ] Add documentation, depending a bit on how the package structure will look like

Frando commented:
Added some docs

RangerMauve commented:
This is looking great, thanks!

Regarding sdk-hyperspace and sdk-native, I think I'd feel happiest about something that looks like this:

```js
const SDK = require('dat-sdk')

// Use the 'native' backend by default
const sdk = await SDK()

// Use the hyperspace backend if you require it explicitly
const hyperspaceBackend = require('dat-sdk/hyperspace')
const sdk = await SDK({backend: hyperspaceBackend})
```

My main reasoning is that the `native` backend works on both web and node without much hassle and is a decent default because of this, then the hyperspace backend can be added when the application author knows what they're doing.

Also, do you think you could add a section to the readme called `## Hyperspace` (somewhere after the build info but before the API docs) which talks about what hyperspace is and when you would want to use it along with the example code for loading it?

calm-rad commented:
Apologies for the delayed reply!

Is there still a way to make a persisting exception for specific hyperdrives/hypercores with this restructuring?

As long as there is some way to achieve that functionality I'm fine with whatever changes allow for more ecosystem interoperability and makes life easier for maintainers.

RangerMauve commented:
@calm-rad If we get rid of the per-drive/core persist flag, then you'll have two alternatives:

- Have a separate swarm which is persist:false for the non-persisted drives (doesn't work for hyperspace anyways)
- Invoke `destroyStorage` after the drive/core isn't used anymore to delete any data it loaded.

Doe either of these approaches work for your use case? I'm not against keeping the persist flag for native backends and noting that it can't be used with the hyperspace backend if you absolutely need it.

calm-rad commented:
@RangerMauve I should be able to make it work with Gateway using those approaches. :)

Frando commented:
I continued a bit on this today! The setup now works as @RangerMauve suggested.

```
const SDK = require('dat-sdk')
const sdk = await SDK() // native backend, works in browser & node
const sdk = await SDK({ backend: 'hyperspace' }) // hyperspace backend
const sdk = await SDK({ backend: require('dat-sdk/hyperspace') }) // same as above

const Client = require('@hyperspace/client')
const client = new Client()
const sdk = await SDK({ hyperspaceClient: client }) // hyperspace backend with client created outside of SDK
```

I also made the test run once between two native SDKs, once between two hyperspace SDKs, and once between a native and a hyperspace SDK.

This looks quite ready from my regard - if you can give it another review @RangerMauve that'd be great :-)

Frando commented:
I rebased on master & force-pushed.

Frando commented:
Oh I also made a change so that the native tests run on a localhost DHT instead of the public one. This made my tests much faster. If you don't want this I'll remove it again :sweat_smile:

Frando commented:
Ah, I forgot about the browser tests. Fixing this now.

Frando commented:
OK, got the browser tests to work again! Moved the files into `/test` because as long as hyperspace doesn't support browsers, we don't want to run these tests in the browser.

Frando commented:
Yeee, with the recent changes by @mafintosh in hyperspace I got it to work in browsers over a websocket connection!

I enabled the hyperspace backend for browsers. Tests pass! It needs a hyperspace server exposing a websocket. I added a script to start test servers for this.

While at it, I fixed the module import structure so that the hyperspace stuff is only imported when needed/wanted (and not always).

I also updated the docs, and added a paragraph on how to run the browser tests.

RangerMauve reviewed (approved):
Thank you so much for your patience! I'll run the tests locally on my machine to get one last check, then get this into a release. 😁

RangerMauve commented:
Out in hyper-sdk@3.0.0. Thank you again for all this work Frando!
